fuzz/quic-server.c

```diff
@@ -0,0 +1,285 @@
+#include "fuzzer.h"
+#include "internal/sockets.h"
+#include "internal/time.h"
+#include "internal/quic_ssl.h"
+
+/* unused, to avoid warning. */
+static int idx;
+
+static OSSL_TIME fake_now;
+
+static OSSL_TIME fake_now_cb(void *arg)
+{
+    return fake_now;
+}
+
+int FuzzerInitialize(int *argc, char ***argv)
+{
+    STACK_OF(SSL_COMP) *comp_methods;
+
+    FuzzerSetRand();
+    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS | OPENSSL_INIT_ASYNC, NULL);
+    OPENSSL_init_ssl(OPENSSL_INIT_LOAD_SSL_STRINGS, NULL);
+    ERR_clear_error();
+    CRYPTO_free_ex_index(0, -1);
+    idx = SSL_get_ex_data_X509_STORE_CTX_idx();
+    comp_methods = SSL_COMP_get_compression_methods();
+    if (comp_methods != NULL)
+        sk_SSL_COMP_sort(comp_methods);
+
+    return 1;
+}
+
+#define HANDSHAKING      0
+#define READING          1
+#define WRITING          2
+#define ACCEPTING_STREAM 3
+#define CREATING_STREAM  4
+#define SWAPPING_STREAM  5
+
+/*
+ * This callback validates and negotiates the desired ALPN on the server side.
+ * Accept any ALPN.
+ */
+static int select_alpn(SSL *ssl, const unsigned char **out,
+                       unsigned char *out_len, const unsigned char *in,
+                       unsigned int in_len, void *arg)
+{
+    return SSL_TLSEXT_ERR_OK;
+}
+
+int FuzzerTestOneInput(const uint8_t *buf, size_t len)
+{
+    SSL *server = NULL, *stream = NULL;
+    SSL *allstreams[] = {NULL, NULL, NULL, NULL};
+    size_t i, thisstream = 0, numstreams = 1;
+    BIO *in;
+    BIO *out;
+    SSL_CTX *ctx;
+    BIO_ADDR *peer_addr = NULL;
+    struct in_addr ina = {0};
+    struct timeval tv;
+    int state = HANDSHAKING;
+    uint8_t tmp[1024];
+    int writelen = 0;
+
+    if (len == 0)
+        return 0;
+
+    /* This only fuzzes the initial flow from the server so far. */
+    ctx = SSL_CTX_new(OSSL_QUIC_server_method());
+    if (ctx == NULL)
+        goto end;
+
+    SSL_CTX_set_alpn_select_cb(ctx, select_alpn, NULL);
+
+    // server = SSL_new(ctx);
+    server = SSL_new_listener(ctx, 0);
+    allstreams[0] = stream = server;
+    if (server == NULL)
+        goto end;
+
+    fake_now = ossl_ms2time(1);
+    if (!ossl_quic_set_override_now_cb(server, fake_now_cb, NULL))
+        goto end;
+
+    peer_addr = BIO_ADDR_new();
+    if (peer_addr == NULL)
+        goto end;
+
+    ina.s_addr = htonl(0x7f000001UL);
+
+    if (!BIO_ADDR_rawmake(peer_addr, AF_INET, &ina, sizeof(ina), htons(4433)))
+       goto end;
+
+    in = BIO_new(BIO_s_dgram_mem());
+    if (in == NULL)
+        goto end;
+    out = BIO_new(BIO_s_dgram_mem());
+    if (out == NULL) {
+        BIO_free(in);
+        goto end;
+    }
+    if (!BIO_dgram_set_caps(out, BIO_DGRAM_CAP_HANDLES_DST_ADDR)) {
+        BIO_free(in);
+        BIO_free(out);
+        goto end;
+    }
+    SSL_set_bio(server, in, out);
+    if (SSL_set1_initial_peer_addr(client, peer_addr) != 1)
+        goto end;
+    SSL_set_accept_state(server);
+
+    fprintf(stderr, "ANDREW: 0\n");
```

Comment:
I assume this is just to see if it gets there, and that this will be removed before merging?

Reply:
Yes, was doing some debugging and will clean it up when this is ready